### ISSUE TITLE
PVP - versioning

### ISSUE THREAD

**Issue by phadej** [aurapm/haskell-versions#3]
Looks like that your package(s) are following the Semantic Versioning -scheme. That's great, but it would be even better if it followed the [Package Versioning Policy (PVP)](https://wiki.haskell.org/Package_versioning_policy). In a nutshell, PVP has the same principles as the SemVer, but it talks specifically about Haskell, and has different meaning for the version digits: *major1.major2.minor.patch*. Even that bumping the versions according do SemVer doesn't violate PVP, it is unnecessarily over-conservative, and thus might be inconvenient for the downstream users. For example [Stackage LTS](https://github.com/fpco/lts-haskell#how-it-works) allow only the *minor* (the third digit) updates inside major LTS version.

--- 

From [the homepage of Hackage](http://hackage.haskell.org/)

> Guidelines for Hackage Packages:
> 
> - All packages should follow the [Package Versioning Policy (PVP)](https://wiki.haskell.org/Package_versioning_policy).

---

cc @bergmark @hvr

**Comment by fosskers** [comment]
I've always found pvp to be overly liberal and awkward to increment. I'm also curious how semver could be inconvenient for users. Have there been documented cases of this?

**Comment by hvr** [comment]
@fosskers the inconvenience is that everyone (including the tooling) expects PVP-version structure and sets versions bounds accordingly. If a package violates this assumption it results incorrect versions bounds which thwarts the primary purpose the PVP was created for.

**Comment by fosskers** [comment]
PVP is clearer to me after reading the page linked by @phadej . That said, having two numbers for major version (*A.B*) seems superfluous, and I claim through anecdote that many packages sit needlessly at `0.x` for their lifetimes. If that's true, then let's just do away with the leading 0.

PVP has three incrementation rules. To directly compare PVP to SemVer (MAJOR.MINOR.PATCH), we have:

1. Increment *A.B* if there is a breaking API change (SemVer: Increment MAJOR)
2. Increment *C* is there is a non-breaking API change (SemVer: Increment MINOR)
3. Otherwise, increment the other components in any way (SemVer: Increment PATCH or pre-release)

From the PVP spec:

> A package version number should have the form A.B.C, and may optionally have any number of additional components...

Meaning a package could have any number of integer (stated later) components over 3. Does Stackage expect exactly 4?

If tools only need to assume 3, is there real harm in adhering to SemVer? I see that `versions` was successfully added to today's nightly, despite being SemVer.

**Comment by bergmark** [comment]
Related comment: https://github.com/jswebtools/language-ecmascript/issues/60#issuecomment-70073959

**Comment by fosskers** [comment]
A casual glance of [the current LTS index](https://www.stackage.org/lts-5.14) shows a non-trivial number of packages not following PVP as defined here and in the linked document above. I see some with just *A.B* and no *C*, even.

**Comment by bergmark** [comment]
The literal number of components does not signify which version scheme is followed, it's perfectly fine to go 1 -> 1.1 -> 1.1.1 -> 1.1.1.1 (except for complicating usages of the MIN_VERSION macros). 

And since we are mentioning LTS, any package following semver would stop being upgraded in LTS as soon as they make a non-breaking api change since LTS also assumes that packages follow the PVP.

**Comment by fosskers** [comment]
> any package following semver would stop being upgraded in LTS as soon as they make a non-breaking api change since LTS also assumes that packages follow the PVP

True, but that might not be the end of the world. People using LTS realize they're not getting the bleeding-edge, and they have to live with that.

I would switch to PVP in a heart beat if I could be convinced it was demonstrably superior to SemVer.

**Comment by hvr** [comment]
@fosskers It's not relevant whether the PVP is superior to SemVer or vice-versa for the matter at hand: Hackage uploaders are asked to follow the common guidelines, and those clearly state that uploaded packages are expected to follow the PVP. You don't have to like the conventions, but you are nevertheless expected to follow them for the sake of the community. When in rome, do as the romans do... :-)

**Comment by mightybyte** [comment]
> I would switch to PVP in a heart beat if I could be convinced it was demonstrably superior to SemVer.

The PVP is a superset of SemVer.  Make your SemVer version, then concatenate the prefix "0." (or "1." or whatever makes you happy) onto the front.  The important point though is that most of the value of both PVP and SemVer comes from network effects.  It's a communication standard between developers. And it just so happens that in the Haskell community the PVP is the standard.

**Comment by fosskers** [comment]
> The PVP is a superset of SemVer

Not entirely true, as SemVer has semantics for rerelease and metadata components, but I see what you mean.

**Comment by fosskers** [comment]
Thanks for the feedback everyone, I'm pondering my course of action.

**Comment by mightybyte** [comment]
> Not entirely true, as SemVer has semantics for rerelease and metadata components, but I see what you mean.

@fosskers Yes, you're right.  But I believe Cabal files do not support SemVer's pre-release and metadata components, so they shouldn't be a part of the discussion in this context.

Also, your package description says that SemVer "has the best constraints on comparisons".  I would argue that in the presence of a pre-release version, SemVer's comparison method is more complex and less intuitive than that of Cabal PVP-style versions.

But again, these are just side notes to the most important point that most Haskell package authors expect the PVP.
